File: LeetCode/67-add-binary/Solution3.java
import java.math.BigInteger;

class Solution3 {
  /**
   * Adds two binary strings using bit manipulation and returns their sum as a binary string.
   *
   * @param a The first binary string.
   * @param b The second binary string.
   * @return The sum of the two binary strings as a binary string.
   */
  public String addBinary(String a, String b) {
    // Convert binary strings to BigInteger objects
    BigInteger x = new BigInteger(a, 2);
    BigInteger y = new BigInteger(b, 2);

    // Define a BigInteger object representing zero
    BigInteger zero = new BigInteger("0", 2);

    BigInteger carry, answer;

    // Loop until there is no carry left
    while (y.compareTo(zero) != 0) {
      // XOR the two numbers to get the sum without carry
      answer = x.xor(y);

      // AND the two numbers and shift left to get the carry
      carry = x.and(y).shiftLeft(1);

      // Update x to the current answer
      x = answer;

      // Update y to the current carry
      y = carry;
    }

    // Convert the result back to a binary string and return it
    return x.toString(2);
  }
}

File: LeetCode/67-add-binary/solution3.ts
/**
 * Adds two binary strings and returns their sum as a binary string.
 *
 * @param {string} a - The first binary string.
 * @param {string} b - The second binary string.
 * @returns {string} - The sum of the two binary strings as a binary string.
 */
function addBinary(a: string, b: string): string {
  // Convert binary strings to BigInt
  let x = BigInt(`0b${a}`);
  let y = BigInt(`0b${b}`);

  // Loop until there is no carry
  while (y !== 0n) {
    // Current answer without carry is XOR of x and y
    const answer = x ^ y;
    // Current carry is left-shifted AND of x and y
    const carry = (x & y) << 1n;
    // Prepare for the next loop
    x = answer;
    y = carry;
  }

  // Convert the final result back to a binary string
  return x.toString(2);
};
